### PR TITLE
Fix compatibility to dbal 2.13 by defining the statement class

### LINKED ISSUE #112: Incompatibility to current doctrine/dbal 2.13.*
It seems like that the crate connection does not implement the dbal `Connection` interface and so currently failing. For dbal 2.13.* (2.12.* works). See comment by @mdumoulin here https://github.com/doctrine/dbal/issues/4619#issuecomment-827900550.

EDIT: Seems like correct interface is used  over the [ServerInfoAwareInterface](https://github.com/crate/crate-dbal/blob/751f3b78f949c97eda15adc0254308e9d7944eed/src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php#L28) but the false statement class is set for PDO.

### PR DESCRIPTION
**Pull request by alexander-schranz**
## Summary of the changes / Why this is an improvement

Fixes #112

## Checklist

 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

File: src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php [hunk 1]
```diff
@@ -36,6 +37,7 @@ class PDOConnection extends PDO implements ServerInfoAwareConnection
     public function __construct($dsn, $user = null, $password = null, array $options = null)
     {
         parent::__construct($dsn, $user, $password, $options);
+        $this->setAttribute(PDO::ATTR_STATEMENT_CLASS, [Statement::class, []]);
```
Comment (alexander-schranz) [inline src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php:40]: not sure if crate supports this. See also #112 and https://github.com/doctrine/dbal/issues/4619#issuecomment-827900550

**Comment by alexander-schranz** [comment]
Looks like the CI is failing with:

> Crate\PDO\Exception\PDOException: Unsupported driver attribute

Open for any input here what the best would be to create the correct statement here.

**Comment by amotl** [comment]
Hi Alexander,

thank you for this patch. Currently, CI is still installing `doctrine/dbal (2.10.4)`.

Can I ask you to accompany your patch by a respective bump to the Doctrine DBAL version used? Then, CI will pick it up run the corresponding tests on it [1].

With kind regards,
Andreas.

[1] Please note that `composer update --lock` needs to be invoked with PHP 7.2, as this currently is the minimum supported version before moving on to PHP 8.0, which will implicitly drop support for PHP 7.2 again. Please let me know if your system/sandbox is not able to do that, so I will be able to chime in.

File: src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php (hunk 1 above)
Comment (amotl) [reply src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php:40]: [1] shows lots of `Crate\PDO\Exception\PDOException: Unsupported driver attribute` errors. Can you make sense of that and evaluate whether corresponding support can be added to `crate-pdo`?

[1] https://github.com/crate/crate-dbal/runs/2452374425#step:9:17

**Comment by alexander-schranz** [comment]
@amotl did update the lock file.

File: src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php (hunk 1 above)
Comment (alexander-schranz) [reply src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php:40]: Not 100% sure but I think something like this is required: https://github.com/crate/crate-pdo/pull/121

Comment (amotl) [reply src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php:40]: Thank you for https://github.com/crate/crate-pdo/pull/121. It has just converged into [`crate-pdo==2.1.0`](https://packagist.org/packages/crate/crate-pdo#2.1.0).

Comment (alexander-schranz) [reply src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php:40]: Thank you did remove the `composer.lock` here also. As we will now need to require `2.1` I think we will run into problems with `7.2`  as the crate-pdo 2.1.0 does not support `7.2` how should we handle this?

File: composer.json [hunk 2]
```diff
@@ -17,7 +17,7 @@
     "require": {
         "php": "^7.2",
```
Comment (amotl) [inline composer.json:18]:
```suggestion
        "php": "^7.3",
```

Comment (amotl) [reply composer.json]: Also, don't forget to also adjust the GHA workflow file as seen in #114.

File: src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php (hunk 1 above)
Comment (amotl) [reply src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php:40]: Isn't PHP 7.2 end of life? We already dropped support on this with crate-pdo 2.1.0 on purpose. So, let's also carry this forward here?

Comment (alexander-schranz) [reply src/Crate/DBAL/Driver/PDOCrate/PDOConnection.php:40]: yes 7.2 is end of life. I'm fine with dropping it here also https://www.php.net/supported-versions.php

**Comment by alexander-schranz** [comment]
@amotl looks like the `statementClass` attribute was never implemented I did update the code that it is now used. https://github.com/crate/crate-pdo/pull/123

Locally with the PR I still get 3 failing tests but not sure if they only fail for me:


<details>
<summary>
Errors:
</summary>

```
PHPUnit 8.5.15 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.4.16 with Xdebug 2.9.8
Configuration: /private/tmp/crate-dbal/phpunit.xml.dist

............F..........F.F.S.S.S.SS............................  63 / 217 ( 29%)
..............SSSSSSSSSSSSSSSS................................. 126 / 217 ( 58%)
.................................SS........SSS.........S....... 189 / 217 ( 87%)
.S....S.....SSS...........FF                                    217 / 217 (100%)

Time: 6.68 seconds, Memory: 12.00 MB

There were 5 failures:

1) Crate\Test\DBAL\Functional\DataAccessTestCase::testPrepareWithFetchAll
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     'test_int' => 1
     'test_string' => 'foo'
-    'test_datetime' => 1262340610000
+    'test_datetime' => 1262337010000
     'test_array' => Array (...)
     'test_object' => Array (...)
 )

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Functional/DataAccessTest.php:139

2) Crate\Test\DBAL\Functional\DataAccessTestCase::testExecuteQueryBindDateTimeType
Failed asserting that 0 matches expected 1.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Functional/DataAccessTest.php:313

3) Crate\Test\DBAL\Functional\DataAccessTestCase::testPrepareQueryBindValueDateTimeType
Failed asserting that 0 matches expected 1.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Functional/DataAccessTest.php:348

4) Crate\Test\DBAL\Platforms\CratePlatformTest::testAsciiSQLDeclaration with data set #0 ('VARCHAR(12)', array(12))
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'VARCHAR(12)'
+'TEXT'

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1472

5) Crate\Test\DBAL\Platforms\CratePlatformTest::testAsciiSQLDeclaration with data set #1 ('CHAR(12)', array(12, true))
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'CHAR(12)'
+'TEXT'

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1472

--

There were 32 skipped tests:

1) Crate\Test\DBAL\Functional\DataAccessTestCase::testDateArithmetics
Data add day expression not supported by crate platform

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Functional/DataAccessTest.php:381

2) Crate\Test\DBAL\Functional\DataAccessTestCase::testBitComparisonExpressionSupport
Bit comparison expression not supported by crate

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Functional/DataAccessTest.php:416

3) Crate\Test\DBAL\Functional\DataAccessTestCase::testFetchAllSupportFetchClass
PDO::FETCH_CLASS not supported by crate PDO

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Functional/DataAccessTest.php:479

4) Crate\Test\DBAL\Functional\DataAccessTestCase::testSetFetchModeClassFetchAll
PDO::FETCH_CLASS not supported crate PDO

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Functional/DataAccessTest.php:523

5) Crate\Test\DBAL\Functional\DataAccessTestCase::testSetFetchModeClassFetch
PDO::FETCH_CLASS not supported by crate PDO

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Functional/DataAccessTest.php:545

6) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesForeignKeyCreationSql
Platform does not support FOREIGN KEY constraints.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:79

7) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesDecimalTypeDeclarationSQL with data set #0 (array(), 'NUMERIC(10, 0)')
Platform does not support any decleration of datatype DECIMAL.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:95

8) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesDecimalTypeDeclarationSQL with data set #1 (array(true), 'NUMERIC(10, 0)')
Platform does not support any decleration of datatype DECIMAL.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:95

9) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesDecimalTypeDeclarationSQL with data set #2 (array(false), 'NUMERIC(10, 0)')
Platform does not support any decleration of datatype DECIMAL.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:95

10) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesDecimalTypeDeclarationSQL with data set #3 (array(5), 'NUMERIC(5, 0)')
Platform does not support any decleration of datatype DECIMAL.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:95

11) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesDecimalTypeDeclarationSQL with data set #4 (array(5), 'NUMERIC(10, 5)')
Platform does not support any decleration of datatype DECIMAL.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:95

12) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesDecimalTypeDeclarationSQL with data set #5 (array(8, 2), 'NUMERIC(8, 2)')
Platform does not support any decleration of datatype DECIMAL.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:95

13) Crate\Test\DBAL\Platforms\CratePlatformTest::testAlterTableChangeQuotedColumn
Platform does not support ALTER TABLE.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:107

14) Crate\Test\DBAL\Platforms\CratePlatformTest::testQuotedColumnInForeignKeyPropagation
Platform does not support ADD FOREIGN KEY.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:138

15) Crate\Test\DBAL\Platforms\CratePlatformTest::testQuotesAlterTableRenameColumn
Platform does not support ALTER TABLE.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:158

16) Crate\Test\DBAL\Platforms\CratePlatformTest::testQuotesAlterTableChangeColumnLength
Platform does not support ALTER TABLE.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:168

17) Crate\Test\DBAL\Platforms\CratePlatformTest::testQuotesAlterTableRenameIndexInSchema
Platform does not support ALTER TABLE.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:178

18) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesAlterTableRenameColumnSQL
Platform does not support ALTER TABLE.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:195

19) Crate\Test\DBAL\Platforms\CratePlatformTest::testQuotesTableIdentifiersInAlterTableSQL
Platform does not support ALTER TABLE.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:205

20) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesAlterTableRenameIndexUsedByForeignKeySQL
Platform does not support ALTER TABLE.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:215

21) Crate\Test\DBAL\Platforms\CratePlatformTest::testAlterStringToFixedString
Platform does not support ALTER TABLE.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:225

22) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesIndexCreationSql
Platform does not support CREATE INDEX.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:69
/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:199

23) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesUniqueIndexCreationSql
Platform does not support CREATE UNIQUE INDEX.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:74
/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:211

24) Crate\Test\DBAL\Platforms\CratePlatformTest::testCreateTableColumnComments
Platform does not support Column comments.

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:517
/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:474

25) Crate\Test\DBAL\Platforms\CratePlatformTest::testAlterTableColumnComments
Platform does not support Column comments.

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:525
/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:499

26) Crate\Test\DBAL\Platforms\CratePlatformTest::testCreateTableColumnTypeComments
Platform does not support Column comments.

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:533
/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:509

27) Crate\Test\DBAL\Platforms\CratePlatformTest::testQuotesReservedKeywordInTruncateTableSQL
Platform does not support TRUNCATE TABLE.

/private/tmp/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:496
/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:730

28) Crate\Test\DBAL\Platforms\CratePlatformTest::testReturnsBinaryTypeLongerThanMaxDeclarationSQL
Not applicable to the platform

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:835

29) Crate\Test\DBAL\Platforms\CratePlatformTest::testQuotesDropForeignKeySQL
Crate\DBAL\Platforms\CratePlatform4 does not support foreign key constraints.

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1074

30) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesInlineColumnCommentSQL with data set "regular comment" ('Regular comment', 'COMMENT 'Regular comment'')
Crate\DBAL\Platforms\CratePlatform4 does not support inline column comments.

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1173

31) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesInlineColumnCommentSQL with data set "comment requiring escaping" ('Using inline comment delimiter ' works', 'COMMENT 'Using inline comment...works'')
Crate\DBAL\Platforms\CratePlatform4 does not support inline column comments.

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1173

32) Crate\Test\DBAL\Platforms\CratePlatformTest::testGeneratesInlineColumnCommentSQL with data set "empty comment" ('', 'COMMENT ''')
Crate\DBAL\Platforms\CratePlatform4 does not support inline column comments.

/private/tmp/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1173
```

</details>

**Comment by amotl** [comment]
> Looks like the `statementClass` attribute was never implemented I did update the code that it is now used: https://github.com/crate/crate-pdo/pull/123.

Thank you. I turned that into another release, see https://github.com/crate/crate-pdo/pull/124.

**Comment by amotl** [comment]
> Locally with the PR I still get 3 failing tests but not sure if they only fail for me:

Down to two, see [1].

```php
There were 2 failures:

1) Crate\Test\DBAL\Platforms\CratePlatformTest::testAsciiSQLDeclaration with data set #0 ('VARCHAR(12)', array(12))
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'VARCHAR(12)'
+'TEXT'

/home/runner/work/crate-dbal/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1472

2) Crate\Test\DBAL\Platforms\CratePlatformTest::testAsciiSQLDeclaration with data set #1 ('CHAR(12)', array(12, true))
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'CHAR(12)'
+'TEXT'

/home/runner/work/crate-dbal/crate-dbal/vendor/doctrine/dbal/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1472
```

[1] https://github.com/crate/crate-dbal/runs/2453539267#step:7:17

**Comment by alexander-schranz** [comment]
@amotl Now only 2 tests are failing:

I did have a look at previous runs and seems like this are new tests (first line this run, second line old CI run on github):

```
.S....S.....SSS...........FF
S....S.....SSS...........
```

I will adjust the test case as I think there is in crate no difference between `VARCHAR(12)` or `TEXT` type.

**Comment by amotl** [comment]
I can confirm all tests work fine on my machine with your most recent b11696a.

**Comment by alexander-schranz** [comment]
@amotl  looks good now from my side :)

**Comment by alexander-schranz** [comment]
@amotl Thx for your support! The PHP 8 PR #111 is now also rebased.

**Comment by amotl** [comment]
Thank you!

**Pull request merged**